plugin/arcgrid/src/test/java/it/geosolutions/imageio/plugins/arcgrid/raster/AsciiGridRasterTest.java

```diff
@@ -0,0 +1,107 @@
+package it.geosolutions.imageio.plugins.arcgrid.raster;
+
+
+import java.awt.Point;
+import java.awt.image.BufferedImage;
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintStream;
+
+
+
+import it.geosolutions.imageio.plugins.arcgrid.AsciiGridsImageReader;
+import it.geosolutions.imageio.plugins.arcgrid.spi.AsciiGridsImageReaderSpi;
+import junit.framework.TestCase;
+
+public class AsciiGridRasterTest extends TestCase {
+  public AsciiGridRasterTest(String name) {
+    super(name);
+  }
+
+
+  // needs to be big enough to trigger use of tiles in underlying java imageio code - these
+  // values seem reasonable?
+  private final int width = 1000;
+  private final int height = 1000;
+
+  private String path;
+
+  public void createRaster() throws IOException {
+    File tempFile = File.createTempFile("grid-test", ".asc");
+    path  = tempFile.getAbsolutePath();
+    PrintStream out = new PrintStream(path);
+    out.format("ncols %d\n", width);
+    out.format("nrows %d\n", height);
+    out.format("xllcorner 0\n");
+    out.format("yllcorner 0\n");
+    out.format("cellsize 1\n");
+
+    int counter = 0;
+    for (int x = 0; x < width; x++) {
+      for (int y = 0; y < height; y++) {
+        out.format("%d ", counter);
+        counter++;
+      }
+
+      out.println("");
+    }
+
+    out.close();
+  }
+
+  public void testCanSamplePixelsFromGridInSequence() throws IOException {
+    createRaster();
+    AsciiGridsImageReaderSpi spi = new AsciiGridsImageReaderSpi();
+    AsciiGridsImageReader imageReader = (AsciiGridsImageReader) spi.createReaderInstance(new File(path));
+    imageReader.setInput(new File(path));
+    BufferedImage image = imageReader.read(0);
+
+    final int numPixels = width * height;
+    int pixelCount = 0;
+    while (pixelCount < numPixels) {
+      assertPixelValue(image, pixelCount, pointFromExpectedPixelValue(pixelCount));
+      pixelCount++;
+    }
+  }
+
+  public void testCanSamplePixelsFromGridWithGapsBetweenUnderlyingImageTiles() throws IOException {
+    createRaster();
+    AsciiGridsImageReaderSpi spi = new AsciiGridsImageReaderSpi();
+    AsciiGridsImageReader imageReader = (AsciiGridsImageReader) spi.createReaderInstance(new File(path));
+    imageReader.setInput(new File(path));
+    // in my original test case, i used the geotools gridformat code to construct a coverage and then took the
+    // image out of there.  This would give me an image with 1000x65 tiles which was enough to trigger the issue
+    //  with the tile cache.  Done here, I get a 1000x1000 (no tiling) even though when I step through I can see
+    // the raster reader computing the 1000x65 tile size - any clues?
+    BufferedImage image = imageReader.read(0);
+
+//
+    int tileHeight = image.getTileHeight();
+    if (tileHeight != image.getHeight() && image.getWidth() == image.getTileWidth()) {
```

Comment:
this is failing - the tiling isn't happening, I'm expecting a 65x1000 tile size, but it's not taking effect :/

Reply:
Mimicking the geotools code doesn't seem to help...

```
    final ImageReadParam readP = new ImageReadParam();

    
    final ParameterBlock pbjImageRead = new ParameterBlock();
    // prepare input to handle possible parallelism between different
    // readers
    AsciiGridsImageReaderSpi readerSPI = new AsciiGridsImageReaderSpi();
    pbjImageRead.add(ImageIO.createImageInputStream(new File(path)));
    pbjImageRead.add(0);
    pbjImageRead.add(Boolean.FALSE);
    pbjImageRead.add(Boolean.FALSE);
    pbjImageRead.add(Boolean.FALSE);
    pbjImageRead.add(null);
    pbjImageRead.add(null);
    pbjImageRead.add(readP);
    pbjImageRead.add(readerSPI.createReaderInstance());
    final RenderedOp asciiCoverage = JAI.create("ImageRead", pbjImageRead, new RenderingHints(null));
    PlanarImage image = PlanarImage.wrapRenderedImage(asciiCoverage);
```

It still runs stupidly slow

Reply:
If no-one can spot anything obvious, I'll create a reduce test case and post it back here for eyeballing